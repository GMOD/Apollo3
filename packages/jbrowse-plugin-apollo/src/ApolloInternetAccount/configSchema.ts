import { ConfigurationSchema } from '@jbrowse/core/configuration'
import { BaseInternetAccountConfig } from '@jbrowse/core/pluggableElementTypes'
import { Instance } from 'mobx-state-tree'

const ApolloConfigSchema = ConfigurationSchema(
  'ApolloInternetAccount',
  {
    baseURL: {
      description: 'Location of Apollo server',
      type: 'string',
      defaultValue: '',
    },
    tokenType: {
      description: 'A custom name for a token to include in the header',
      type: 'string',
      defaultValue: 'Bearer',
    },
  },
  { baseConfiguration: BaseInternetAccountConfig, explicitlyTyped: true },
)

export type ApolloInternetAccountConfigModel = typeof ApolloConfigSchema

export type ApolloInternetAccountConfig =
  Instance<ApolloInternetAccountConfigModel>
export default ApolloConfigSchema
