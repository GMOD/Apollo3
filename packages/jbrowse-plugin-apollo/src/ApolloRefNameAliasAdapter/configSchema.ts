import { ConfigurationSchema } from '@jbrowse/core/configuration'

export default ConfigurationSchema(
  'ApolloRefNameAliasAdapter',
  {
    assemblyId: {
      type: 'string',
      defaultValue: '',
    },
  },
  { explicitlyTyped: true },
)
