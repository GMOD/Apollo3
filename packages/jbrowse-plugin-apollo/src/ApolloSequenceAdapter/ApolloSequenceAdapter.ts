import { readConfObject } from '@jbrowse/core/configuration'
import {
  BaseOptions,
  BaseSequenceAdapter,
} from '@jbrowse/core/data_adapters/BaseAdapter'
import { ObservableCreate } from '@jbrowse/core/util/rxjs'
import SimpleFeature, { Feature } from '@jbrowse/core/util/simpleFeature'
import { NoAssemblyRegion, Region } from '@jbrowse/core/util/types'
import { nanoid } from 'nanoid'

import { BackendDriver } from '../BackendDrivers'
import { ApolloSessionModel } from '../session'

// declare global {
//   var rpcServer: import('librpc-web-mod').RpcServer
// }

export interface RefSeq {
  _id: string
  name: string
  description: string
  length: number
}

interface ApolloMessageData {
  apollo: true
  messageId: string
  sequence: string
  regions: Region[]
}

function isApolloMessageData(data?: unknown): data is ApolloMessageData {
  return (
    typeof data === 'object' &&
    data !== null &&
    'apollo' in data &&
    data.apollo === true
  )
}

const isInWebWorker = typeof sessionStorage === 'undefined'

export class ApolloSequenceAdapter extends BaseSequenceAdapter {
  private regions: NoAssemblyRegion[] | undefined

  public async getRefNames(opts: BaseOptions) {
    const regions = await this.getRegions(opts)
    return regions.map((regions) => regions.refName)
  }

  public async getRegions(opts: BaseOptions): Promise<NoAssemblyRegion[]> {
    if (this.regions) {
      return this.regions
    }
    const assemblyId = readConfObject(this.config, 'assemblyId')
    if (!isInWebWorker) {
      const dataStore = (
        this.pluginManager?.rootModel?.session as ApolloSessionModel | undefined
      )?.apolloDataStore
      if (!dataStore) {
        throw new Error('No Apollo data store found')
      }
      const backendDriver = dataStore.getBackendDriver(
        assemblyId,
      ) as BackendDriver
      const regions = await backendDriver.getRegions(assemblyId)
      this.regions = regions
      return regions
    }
    const { signal } = opts
    const regions = await new Promise(
      (
        resolve: (sequence: Region[]) => void,
        reject: (reason: string) => void,
      ) => {
        const timeoutId = setTimeout(() => {
          reject('timeout')
        }, 20_000)
        const messageId = nanoid()
        const messageListener = (event: MessageEvent) => {
          const { data } = event
          if (!isApolloMessageData(data)) {
            return
          }
          if (data.messageId !== messageId) {
            return
          }
          clearTimeout(timeoutId)
          removeEventListener('message', messageListener)
          resolve(data.regions)
        }
        addEventListener('message', messageListener, { signal })
        // @ts-expect-error waiting for types to be published
        globalThis.rpcServer.emit('apollo', {
          apollo: true,
          method: 'getRegions',
          assembly: assemblyId,
          messageId,
        })
      },
    )
    this.regions = regions
    return regions
  }

  /**
   * Fetch features for a certain region
   * @param param -
   * @returns Observable of Feature objects in the region
   */
  public getFeatures(region: Region, opts: BaseOptions) {
    const { signal } = opts
    const { assemblyName, end, refName, start } = region
    return ObservableCreate<Feature>(async (observer) => {
      if (!isInWebWorker) {
        const dataStore = (
          this.pluginManager?.rootModel?.session as
            | ApolloSessionModel
            | undefined
        )?.apolloDataStore
        if (!dataStore) {
          observer.error('No Apollo data store found')
          return
        }
        const backendDriver = dataStore.getBackendDriver(
          assemblyName,
        ) as BackendDriver
        const { seq } = await backendDriver.getSequence(region)
        observer.next(
          new SimpleFeature({
            id: `${refName} ${start}-${end}`,
            data: { refName, start, end, seq },
          }),
        )
        observer.complete()
        return
      }
      const seq = await new Promise(
        (
          resolve: (sequence: string) => void,
          reject: (reason: string) => void,
        ) => {
          const timeoutId = setTimeout(() => {
            reject('timeout')
          }, 20_000)
          const messageId = nanoid()
          const messageListener = (event: MessageEvent) => {
            const { data } = event
            if (!isApolloMessageData(data)) {
              return
            }
            if (data.messageId !== messageId) {
              return
            }
            clearTimeout(timeoutId)
            removeEventListener('message', messageListener)
            resolve(data.sequence)
          }
          addEventListener('message', messageListener, { signal })
          // @ts-expect-error waiting for types to be published
          globalThis.rpcServer.emit('apollo', {
            apollo: true,
            method: 'getSequence',
            region,
            messageId,
          })
        },
      )
      observer.next(
        new SimpleFeature({
          id: `${refName} ${start}-${end}`,
          data: { refName, start, end, seq },
        }),
      )
      observer.complete()
    })
  }

  /**
   * called to provide a hint that data tied to a certain region
   * will not be needed for the foreseeable future and can be purged
   * from caches, etc
   */
  // eslint-disable-next-line @typescript-eslint/no-empty-function
  public freeResources(/* { region } */): void {}
}
