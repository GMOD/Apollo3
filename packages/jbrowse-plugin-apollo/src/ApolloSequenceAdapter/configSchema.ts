import { ConfigurationSchema } from '@jbrowse/core/configuration'

export default ConfigurationSchema(
  'ApolloSequenceAdapter',
  {
    assemblyId: {
      type: 'string',
      defaultValue: '',
    },
  },
  { explicitlyTyped: true },
)
