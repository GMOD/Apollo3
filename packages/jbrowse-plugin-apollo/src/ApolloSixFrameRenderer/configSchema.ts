import { ConfigurationSchema } from '@jbrowse/core/configuration'

export default ConfigurationSchema(
  'ApolloSixFrameRenderer',
  {},
  { explicitlyTyped: true },
)
