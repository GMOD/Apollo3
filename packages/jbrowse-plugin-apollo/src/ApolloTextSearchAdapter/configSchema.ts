import { ConfigurationSchema } from '@jbrowse/core/configuration'

export default ConfigurationSchema(
  'ApolloTextSearchAdapter',
  {
    assemblyNames: {
      type: 'stringArray',
      defaultValue: [],
      description: 'List of assemblies covered by text search adapter',
    },
    trackId: {
      type: 'string',
      defaultValue: '',
    },
    baseURL: {
      type: 'fileLocation',
      defaultValue: {
        uri: '',
        locationType: 'UriLocation',
      },
    },
  },
  { explicitlyTyped: true, explicitIdentifier: 'textSearchAdapterId' },
)
