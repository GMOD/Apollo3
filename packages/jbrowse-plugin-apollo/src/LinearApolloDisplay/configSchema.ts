import { ConfigurationSchema } from '@jbrowse/core/configuration'
import PluginManager from '@jbrowse/core/PluginManager'
import type LinearGenomeViewPlugin from '@jbrowse/plugin-linear-genome-view'

export function configSchemaFactory(pluginManager: PluginManager) {
  const LGVPlugin = pluginManager.getPlugin(
    'LinearGenomeViewPlugin',
  ) as LinearGenomeViewPlugin
  const { baseLinearDisplayConfigSchema } = LGVPlugin.exports

  return ConfigurationSchema(
    'LinearApolloDisplay',
    { height: { type: 'number', defaultValue: 500 } },
    { baseConfiguration: baseLinearDisplayConfigSchema, explicitlyTyped: true },
  )
}
