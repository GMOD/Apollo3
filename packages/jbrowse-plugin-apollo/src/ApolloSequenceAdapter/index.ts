import AdapterType from '@jbrowse/core/pluggableElementTypes/AdapterType'
import PluginManager from '@jbrowse/core/PluginManager'

import { ApolloSequenceAdapter } from './ApolloSequenceAdapter'
import configSchema from './configSchema'

export function installApolloSequenceAdapter(pluginManager: PluginManager) {
  pluginManager.addAdapterType(
    () =>
      new AdapterType({
        name: 'ApolloSequenceAdapter',
        configSchema,
        adapterMetadata: {
          category: null,
          hiddenFromGUI: true,
          displayName: null,
          description: null,
        },
        AdapterClass: ApolloSequenceAdapter,
      }),
  )
}
