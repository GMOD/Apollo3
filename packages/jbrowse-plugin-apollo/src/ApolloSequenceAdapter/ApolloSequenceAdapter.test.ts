import { beforeAll, beforeEach, describe, expect, it } from '@jest/globals'
import fetchMock from 'jest-fetch-mock'
import { toArray } from 'rxjs/operators'

import { ApolloSequenceAdapter, RefSeq } from './ApolloSequenceAdapter'
import configSchema from './configSchema'

const mockRefSeqs: RefSeq[] = [
  {
    _id: '6317dac51a89f156b2e21a70',
    name: 'ctgA',
    description: 'the first contig',
    length: 10000,
  },
  {
    _id: '6317dac51a89f156b2e21b7c',
    name: 'ctgB',
    description: 'the second contig',
    length: 10000,
  },
]

describe('ApolloSequenceAdapter', () => {
  let adapter: ApolloSequenceAdapter
  beforeAll(async () => {
    adapter = new ApolloSequenceAdapter(
      configSchema.create({
        assemblyId: '6317d5436061de774b43e9d6',
        baseURL: { uri: 'http://fake.url', locationType: 'UriLocation' },
      }),
    )
    // populate refSeqs cache before the tests are run
    fetchMock.mockResponseOnce(JSON.stringify(mockRefSeqs))
    await adapter.getRefNames({})
  })
  beforeEach(() => {
    fetchMock.resetMocks()
  })
  it('can get refNames and regions', async () => {
    const refNames = await adapter.getRefNames({})
    expect(refNames).toEqual(mockRefSeqs.map((r) => r.name))
    const regions = await adapter.getRegions({})
    expect(regions).toEqual(
      mockRefSeqs.map((r) => ({ refName: r.name, start: 0, end: r.length })),
    )
  })
  it('can get features', async () => {
    fetchMock.mockResponseOnce('GCGTGCAACAGACTTTCCATGATGCGAGCT')
    const features = adapter.getFeatures(
      { refName: 'ctgA', start: 0, end: 30 },
      {},
    )
    const featuresArray = await features.pipe(toArray()).toPromise()
    expect(featuresArray).toMatchInlineSnapshot(`
      [
        {
          "end": 30,
          "refName": "ctgA",
          "seq": "GCGTGCAACAGACTTTCCATGATGCGAGCT",
          "start": 0,
          "uniqueId": "ctgA 0-30",
        },
      ]
    `)
  })
})
