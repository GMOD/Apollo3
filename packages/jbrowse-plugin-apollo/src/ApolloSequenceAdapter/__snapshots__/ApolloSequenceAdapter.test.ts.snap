// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApolloSequenceAdapter can get features 1`] = `
[
  {
    "end": 30,
    "refName": "ctgA",
    "seq": "GCGTGCAACAGACTTTCCATGATGCGAGCT",
    "start": 0,
    "uniqueId": "ctgA 0-30",
  },
]
`;
